apple/internal/docc.bzl

```diff
@@ -0,0 +1,133 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines rules for building Apple DocC targets."""
+
+load(
+    "@build_bazel_apple_support//lib:apple_support.bzl",
+    "apple_support",
+)
+load(
+    "@bazel_skylib//lib:dicts.bzl",
+    "dicts",
+)
+load(
+    "//apple:providers.bzl",
+    "DocCBundleInfo",
+)
+load(
+    "//apple/internal/aspects:docc_bundle_info_aspect.bzl",
+    "docc_bundle_info_aspect",
+)
+
+def _docc_archive_impl(ctx):
+    """Builds a .doccarchive for the given module.
+    """
+
+    fallback_bundle_identifier = ctx.attr.fallback_bundle_identifier
+    fallback_bundle_version = ctx.attr.fallback_bundle_version
+    fallback_display_name = ctx.attr.fallback_display_name
+    transform_for_static_hosting = ctx.attr.transform_for_static_hosting
+    docc_info = ctx.attr.dep[0][DocCBundleInfo]  # is a single target but transition forces a list
+
+    if not docc_info.symbol_graph:
+        fail("Expected a symbol graph to be provided in DocCBundleInfo provider for target: %s" % ctx.attr.name)
+
+    if not docc_info.bundle:
+        # TODO: We should generate a default bundle here maybe?
+        fail("Expected a DocC bundle to be provided in DocCBundleInfo provider for target: %s" % ctx.attr.name)
+
+    if ctx.attr.name.endswith(".doccarchive"):
+        doccarchive_dir = ctx.actions.declare_directory(ctx.attr.name)
+    else:
+        doccarchive_dir = ctx.actions.declare_directory("%s.doccarchive" % ctx.attr.name)
+
+    argumets = ctx.actions.args()
+    argumets.add("docc")
+    argumets.add("convert")
+    argumets.add("--index")
+    argumets.add("--fallback-display-name", fallback_display_name)
+    argumets.add("--fallback-bundle-identifier", fallback_bundle_identifier)
+    argumets.add("--fallback-bundle-version", fallback_bundle_version)
+    argumets.add("--output-dir", doccarchive_dir.path)
+    argumets.add("--additional-symbol-graph-dir", docc_info.symbol_graph.path)
+
+    if transform_for_static_hosting:
+        argumets.add("--transform-for-static-hosting")
+
+    # Build and convert the target and symbol graph into a .doccarchive.
+    apple_support.run(
+        ctx,
+        inputs = [docc_info.symbol_graph],
+        outputs = [doccarchive_dir],
+        mnemonic = "DoccArchive",
+        executable = "/usr/bin/xcrun",
+        arguments = [argumets],
+        progress_message = "Building .doccarchive for %{label}",
+    )
+
+    # Create an executable shell script that runs `docc preview` on the .doccarchive.
+    preview_script = ctx.actions.declare_file("%s_preview.sh" % ctx.attr.name)
+    ctx.actions.write(
+        output = preview_script,
+        content = """
+#!/bin/bash
+set -euo pipefail
+""",
```

Comment:
Should there be a call to `/usr/bin/xcrun docc preview ...` in here?

Reply:
Added a preview template and `bazel run` now spins up a local server for the docs